### PR TITLE
feat: add generic display of properties

### PR DESCRIPTION
**Pull request by Bronstrom**
This creates a generic OSCALProperties component and adds it to back matter, controls, and metadata.

Closes https://github.com/EasyDynamics/oscal-react-library/issues/718

### CODE REVIEW

File: packages/oscal-react-library/src/components/OSCALControl.js [hunk 1]
```diff
@@ -44,6 +45,15 @@ function ControlsList(props) {
   } = props;
   return (
     <div>
+      <OSCALProperties
+        properties={control.props}
+        title={
+          <>
+            <OSCALControlLabel id={control.id} label={control.label} component="span" />
+            {` ${control.title} Discussion`}
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALControl.js:53]:
```suggestion
            {` ${control.title} Properties`}
```

?

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      >
+        {children}
+      </Dialog>
+    </>
+  );
+}
+
+function isNistNamespace(namespace: string) {
+  return !namespace || namespace.includes("nist.gov");
+}
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:71]: Consider trying to use something from `OSCALPropUtils` (exporting something that exists or adding a new function)

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    ?.filter((property: any) => !isNistNamespace(property?.ns))
+    .map((property: any) => {
+      items.push(property?.ns);
+    });
+
+  return items;
+}
+
+function OSCALPropertiesTable(props: any) {
+  const { properties, namespace } = props;
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:86]:
```suggestion

interface OSCALPropertiesTableProps {
  properties: Property[]; // Import from "@easydynamics/oscal-types"
  namespace?: string;
}

function OSCALPropertiesTable(props: any) {
  const { properties, namespace } = props;
```

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+            <TableRow>
+              <StyledHeaderTableCell>Name</StyledHeaderTableCell>
+              <StyledHeaderTableCell>Class</StyledHeaderTableCell>
+              <StyledHeaderTableCell>Value</StyledHeaderTableCell>
+            </TableRow>
+          </StyledTableHead>
+          <TableBody>
+            {properties
+              ?.filter((property: any) =>
+                namespace === "NIST OSCAL"
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:105]: Make this a constant if we're going to reuse it -- especially bc i don't love this value but I can't think of something better right now.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        <Stack direction="column">{title} Properties</Stack>
+      </DialogTitle>
+      {/* Handle NIST properties */}
+      <OSCALPropertiesTable properties={properties} namespace={"NIST OSCAL"} />
+      {
+        /* Handle 3rd party properties */
+        thirdPartyNamespaces
+          ?.filter((namespace) => namespace !== "")
+          .map((namespace: any) => (
+            <OSCALPropertiesTable properties={properties} namespace={namespace} />
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:141]:
```suggestion
            <OSCALPropertiesTable properties={properties} namespace={namespace} key={namespace} />
```

**Review by kylelaker — COMMENTED** [review]
Very early thoughts. There are a few design things not written here that we should talk through. I will try to summarize those and provide more info soon.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 2 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx]: This is a little different than what is in `OSCALPropUtils`, more of a validation check rather than providing a NIST namespace. I've gone ahead and added this to `OSCALPropUtils`.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 4 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx]: I agree, my plan was to check-in on what our NIST namespace should be called.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+   * Title of the element the properties resides under
+   */
+  title: string;
+  /**
+   * Children to display in the modal
+   */
+  children: React.ReactNode;
+}
+
+const OSCALPropertiesDialog = (props: OSCALPropertiesDialogProps): ReactElement => {
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:42]:
```suggestion
const OSCALPropertiesDialog: React.FC<OSCALPropertiesDialogProps> = (props) => {
```

File: packages/oscal-react-library/src/components/oscal-utils/OSCALPropUtils.ts [hunk 7]
```diff
@@ -5,6 +5,7 @@ import type { Property } from "@easydynamics/oscal-types";
  * namespace is default namespace.
  */
 const NIST_DEFAULT_NAMESPACE = "https://csrc.nist.gov/ns/oscal";
+export const NIST_NAMESPACE_DOMAIN = "nist.gov";
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/oscal-utils/OSCALPropUtils.ts:8]: I don't want to make assumptions based on domain name -- let's remove this

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+      <OSCALPropertiesTable properties={properties} namespace={NIST_NAMESPACE_DOMAIN} key={NIST_NAMESPACE_DOMAIN} />
+      {
+        /* Handle 3rd party properties */
+        thirdPartyNamespaces
+          ?.filter((namespace) => namespace !== "")
+          .map((namespace: any) => (
+            <OSCALPropertiesTable properties={properties} namespace={namespace} key={namespace} />
+          ))
+      }
+    </OSCALPropertiesDialog>
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:183]: Let's try to sort to put the `https://csrc.nist.gov/ns/oscal` namespace at the top; however, otherwise it should be treated the exact same as other namespaces. Including using the _value_ (not some other string) in the display.

**Review by tuckerzp — APPROVED** [review]
Just had a quick look over things as they are. I had a few comments. 

I also am not sure I understand why `namespaceOf` can't be used instead of `isNistNamespace`. I didn't have a ton of time to think it over, so maybe I am missing something.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+   * Contains properties elements
+   */
+  properties: any;
+  /**
+   * Title of the element the properties resides under
+   */
+  title: string;
+}
+
+export const OSCALProperties = (props: OSCALPropertiesProps) => {
```
[the remark this answers is not in the document]
Comment (tuckerzp) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:166]: To clarify a bit, I think it is best to make things that are objects from OSCAL (like Properties) be the simplest possible component so in theory it can be used elsewhere.

**Comment by Bronstrom** [comment]
> Just had a quick look over things as they are. I had a few comments.
> 
> I also am not sure I understand why `namespaceOf` can't be used instead of `isNistNamespace`. I didn't have a ton of time to think it over, so maybe I am missing something.

`namespaceOf` returns a namespace or the default NIST namespace when there is no namespace provided. `isNistNamespace` validates the namespace being a NIST namespace, which is true when `ns` is empty or an empty string. This validation check is done a few times within `OSCALProperties` and I thought it made sense to be a function. This could be turned into `namespaceOf(ns) === NIST_DEFAULT_NAMESPACE` if we wanted to use that function inside of this one.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+   */
+  namespace?: string;
+}
+
+const OSCALPropertiesTable = (props: OSCALPropertiesTableProps): ReactElement => {
+  const { properties, namespace } = props;
+
+  return (
+    <>
+      <OSCALSystemImplementationTableTitle variant="h6" id={`${namespace}-table-title`}>
```
[the remark this answers is not in the document]
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:120]: I agree, I might hold off on this to the Global Styles implementation. Unless this feel urgent.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:166]: This may make more sense now with there no longer being a default export, which would have followed the naming scheme of the file. Although, what we should also think about is the sheer amount of data that will appear on screen to avoid overwhelming the user - the test catalog I've attached above may give a better idea of why we would want to commonly separate properties into a modal rather than displaying long screens of these properties (especially if there are many 3rd party namespaces in a `props` list). Are there any places in particular you would picture the table existing over just the button?

[the remark this answers is not in the document]
Comment (tuckerzp) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:166]: I don't see a use of it currently, but I think it would be annyoing to switch this around if there ever is a use case. 

Take [OSCAL Revisions](https://github.com/EasyDynamics/oscal-react-library/blob/develop/packages/oscal-react-library/src/components/OSCALRevision.tsx) as an example. Today, we do not actually use the `OSCALRevisions` component outside of `OSCALRevisionsButton`. I do not think the dialog button necessarily encapsulates a list of revisions, it just allows for you to view Revisons. 

I would love to see a similar pattern here. I think making a very basic component that we can then provide wrappers to is a more maintainable pattern overall. 

I was thinking you could do something like:

- `OSCALPropertiesDialog` as the dialog that opens the properties table
- `OSCALProperties` as a table
- `OSCALProperty` as a row within that table

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+  properties
+    ?.filter((property: any) => !isNistNamespace(property?.ns))
+    .map((property: any) => {
+      if (!items.includes(property?.ns)) {
+        items.push(property?.ns);
+      }
+    });
+
+  return items;
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:49]:
```suggestion
  return [...new Set(
    properties
      ?.filter((prop) => namespaceOf(property.ns) !== NIST_DEFAULT_NAMESPACE)
      .map((prop) => prop.ns)
      .filter((prop): prop is string => !!prop.ns)
  )]
```

File: packages/oscal-react-library/src/components/oscal-utils/OSCALPropUtils.ts [hunk 12]
```diff
@@ -16,6 +16,16 @@ export function namespaceOf(ns: string | undefined): string {
...
+/**
+ * Validates if the namespace is a NIST namespace.
+ *
+ * @param namespace the namespace
+ * @returns a boolean describing whether the namespace is
+ */
+export function isNistNamespace(ns: string | undefined): boolean {
+  return !ns || ns === NIST_DEFAULT_NAMESPACE;
+}
+
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/oscal-utils/OSCALPropUtils.ts:28]:
```suggestion
```

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+          <TableBody>
+            {properties
+              ?.filter((property: any) =>
+                namespace === NIST_DEFAULT_NAMESPACE
+                  ? isNistNamespace(property?.ns)
+                  : property?.ns === namespace
+              )
+              .map((property: any) => (
+                <OSCALProperty property={property} />
+              ))}
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:111]:
```suggestion
              ?.filter((property) =>
                namespace === NIST_DEFAULT_NAMESPACE
                  ? isNistNamespace(property?.ns)
                  : property?.ns === namespace
              )
              .map((property) => (
                <OSCALProperty property={property} />
              ))}
```

**Comment by tuckerzp** [comment]
I discussed this briefly with @Bronstrom, but I wonder if filtering out non nist namespaces to a separate list to only then put it in more or less the same list is the best approach.

To me, this feels more like we need to `sort`. Anything that is considered to be in the default namespace (found preferably using `namespaceOf` and a boolean check) would be put at the top of the list while everything else can just remain in the same order. 

I honestly am unsure if this cleans up the code a ton, but conceptually it makes more sense to me in terms of what we are actually trying to do. I also think it will make the return statement of the component a lot cleaner.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 14]
```diff
@@ -30,21 +30,40 @@ const OSCALPropertiesButton = styled(Button)(
 );
 
 /**
- * Accumulates all instances of third-party namespaces into a list.
+ * Sorts the namespaces by NIST namespaces first, orders third-party namespaces alphabetically,
+ * then orders by name.
  *
  * @param {Property[]} properties A list of properties
- * @returns {string[]} A list of third-party namespaces
+ * @returns {Property[] | undefined} A list of third-party namespaces
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx]: Just move the types to TS and remove from JSDoc

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 15]
```diff
@@ -30,21 +30,40 @@ const OSCALPropertiesButton = styled(Button)(
...
-    .map((property: any) => {
-      if (!items.includes(property?.ns)) {
-        items.push(property?.ns);
-      }
-    });
+  properties?.map((property: any) => {
+    if (!items.includes(namespaceOf(property?.ns))) {
+      items.push(namespaceOf(property?.ns));
+    }
+  });
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:66]: If you're going to do this over the suggestion I had made, then this needs to be `forEach` not a `map`. You're `map`ing to void/undefined. Which is very ungood.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 16]
```diff
@@ -101,11 +120,7 @@ const OSCALProperties = (props: OSCALPropertiesProps): ReactElement => {
           </StyledTableHead>
           <TableBody>
             {properties
-              ?.filter((property: any) =>
-                namespace === NIST_DEFAULT_NAMESPACE
-                  ? isNistNamespace(property?.ns)
-                  : property?.ns === namespace
-              )
+              ?.filter((property: any) => namespaceOf(property?.ns) === namespace)
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:123]: I do not want any `: any`s as part of this -- we know these are `Property`s and we should make sure that is able to inferred in all these places. If it can't we're doing something wrong somewhere else.

There are a lot of these so I am just going to comment it here instead of everywhere.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 17]
```diff
@@ -30,21 +30,40 @@ const OSCALPropertiesButton = styled(Button)(
...
+    namespaceOf(propA.ns) > namespaceOf(propB.ns) ? 1 : -1
+  );
+  const nistNamespaceFirstProperties = namespaceSortedProperties?.sort((propA, propB) =>
+    isNistNamespace(propA.ns) && !isNistNamespace(propB.ns) ? 1 : -1
+  );
+  const nameSortedProperties = nistNamespaceFirstProperties?.sort((propA, propB) =>
+    namespaceOf(propA.ns) === namespaceOf(propB.ns) && propA?.name > propB?.name ? 1 : -1
+  );
+
+  return nameSortedProperties;
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:50]: This is... complicated. And I still _really_ do not like `isNistNamespace`'s existence.

**Comment by Bronstrom** [comment]
Yesterday, @tuckerzp and I discussed making some the `filter`s into `sort`s. `sort`s  have proven to look a lot cleaner than using `filter`s, which really should be used if we're removing some of the data rather than ordering it. I've turned the third parties namespace function into one that captures all namespaces in a list, in addition to creating a function to sort properties in the correct order. This ordering is first by the NIST namespace, then 3rd Party extension namespaces, then by name. Three sorts seemed to achieve this well and have been split apart into separate constants for readability.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 17 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:50]: I thought that helped with readability of the sorts. Do you have a better suggestion? Also, didn't realize you don't like `isNistNamespace` - thought it would be helpful to create a reusable function if we're making the validation more than once.

**Comment by kylelaker** [comment]
> `sort`s have proven to look a lot cleaner than using `filter`s, which really should be used if we're removing some of the data rather than ordering it

I think this is a matter of preference not proof 🤣 I think the chained sorting that is required adds complexity and definitely means that we have to test the heck out of that sorting code.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+            properties={sortProperties(nist)}
+            namespace={NIST_DEFAULT_NAMESPACE}
+            key={NIST_DEFAULT_NAMESPACE}
+          />
+          {
+            /* Handle 3rd party properties */
+            Object.entries(thirdParties)
+              .sort(([key1, props1], [key2, props2]) => key1.localeCompare(key2))
+              .map(([key, props]) => (
+                <OSCALProperties properties={sortProperties(props)} namespace={key} key={key} />
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:172]: If it's going to be broken into a different function I'd rather see this as

```ts

function byName(a: Property, b: Property) {
  return a.name.localeCompare(b.name);
}

<OSCALProperties properties={props.sort(byName)} namespace={key} key={key} />
```

This way it's more scriptive _how_ they're being sorted.

File: packages/oscal-react-library/src/components/OSCALBackMatter.tsx [hunk 19]
```diff
@@ -215,6 +215,10 @@ function BackMatterResource(props: BackMatterResourceProps) {
           action={<CitationDisplay resource={resource} />}
         />
         <CardContent>
+          <OSCALPropertiesDialog
+            properties={resource?.props}
+            title={resource?.title ?? resource?.uuid}
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALBackMatter.tsx:220]: Let's not use a UUID in the title... I'd rather see us do something like `<NotSpecifiedTypography component="span">Resource</NotSpecifiedTypography>`

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+
+interface OSCALPropertyProps {
+  /**
+   * Contains property elements
+   */
+  property: Property;
+}
+
+const OSCALProperty = (props: OSCALPropertyProps): ReactElement => {
+  const { property } = props;
```
Comment (tuckerzp) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:52]: Can we update all the components in this file to do this?

```suggestion
const OSCALProperty: React.FC<OSCALPropertyProps> = ({ property }) => { 
```

File: packages/oscal-react-library/src/components/OSCALControl.js [hunk 21]
```diff
@@ -44,6 +45,15 @@ function ControlsList(props) {
...
     <div>
+      <OSCALPropertiesDialog
+        properties={control.props}
+        title={
+          <>
+            <OSCALControlLabel id={control.id} label={control.label} component="span" />
+            {` ${control.title}`}
+          </>
+        }
+      />
```
Comment (tuckerzp) [inline packages/oscal-react-library/src/components/OSCALControl.js:56]: Is there a cleaner way to deal with this? Is there a reason each component using an `OSCALPropertiesDialog` would need to create there own title? Would having this just be the `children` be a cleaner solution?

File: packages/oscal-react-library/src/components/OSCALBackMatter.tsx (hunk 19 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALBackMatter.tsx]: I added a fallback title within `Properties`, maybe we would want to make it more generic within Properties. Maybe it could just be "Not Specified Properties"?

File: packages/oscal-react-library/src/components/OSCALControl.js (hunk 21 above)
Comment (tuckerzp) [reply packages/oscal-react-library/src/components/OSCALControl.js:56]: I personally would like if we stopped putting huge components as a prop. At the very least lets do something like

```TypeScript
const controlLabel =  (
  <>
    <OSCALControlLabel id={control.id} label={control.label} component="span" />
      {` ${control.title}`}
  </>
);

return (
  <OSCALPropertiesDialog title={controlLabel />
);

File: packages/oscal-react-library/src/components/OSCALCatalogGroups.test.tsx [hunk 22]
```diff
@@ -52,6 +52,73 @@ const testGroups: ControlGroup[] = [
   },
 ];
 
+const testGroupswithProps: ControlGroup[] = [
+  {
```
Comment (Bronstrom) [inline packages/oscal-react-library/src/components/OSCALCatalogGroups.test.tsx:56]: Not sure how I feel about this test data here (including the test data before this one). Maybe this could be moved within `test-data`?

**Comment by Bronstrom** [comment]
Here's some screenshots of how the modal is looking:

generic_properties_modal_1

generic_properties_modal_2

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+  return (
+    <StyledTooltip title={property.remarks ?? ""}>
+      <StyledTableRow key={property.uuid}>
+        <TableCell>{property.name ?? NO_INFORMATION}</TableCell>
+        <TableCell>{property.class ?? NO_INFORMATION}</TableCell>
+        <TableCell>{property.value ?? NO_INFORMATION}</TableCell>
+      </StyledTableRow>
+    </StyledTooltip>
+  );
+};
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:63]: I am concerned that our only option to represent a single Property is to have it as a table row... I suppose it's probably fine because I can't think of another place where we may need it...

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+              <StyledHeaderTableCell>Name</StyledHeaderTableCell>
+              <StyledHeaderTableCell>Class</StyledHeaderTableCell>
+              <StyledHeaderTableCell>Value</StyledHeaderTableCell>
+            </TableRow>
+          </StyledTableHead>
+          <TableBody>
+            {properties
+              ?.filter((property) => namespaceOf(property?.ns) === namespace)
+              .map((property) => (
+                <OSCALProperty property={property} key={`$namespaceOf(property?.ns)-properties`} />
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:95]:
```suggestion
                <OSCALProperty property={property} key={`${namespaceOf(property?.ns)}-properties`} />
```

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+  const [open, setOpen] = React.useState(false);
+
+  if (!properties) {
+    return null;
+  }
+  const namespaceToProps: Record<string, Property[]> = {};
+  for (const prop of properties) {
+    namespaceToProps[namespaceOf(prop.ns)] ??= [];
+    namespaceToProps[namespaceOf(prop.ns)].push(prop);
+  }
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:128]:
```suggestion
  for (const prop of properties) {
    const ns = namespaceOf(prop.ns);
    namespaceToProps[ns] ??= [];
    namespaceToProps[ns].push(prop);
  }
```

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 26]
```diff
@@ -0,0 +1,179 @@
...
+            properties={nist?.sort(byName)}
+            namespace={NIST_DEFAULT_NAMESPACE}
+            key={NIST_DEFAULT_NAMESPACE}
+          />
+          {
+            /* Handle 3rd party properties */
+            Object.entries(thirdParties)
+              .sort(([key1], [key2]) => key1.localeCompare(key2))
+              .map(([key, props]) => (
+                <OSCALProperties properties={props?.sort(byName)} namespace={key} key={key} />
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:172]:
```suggestion
                <OSCALProperties properties={props.sort(byName)} namespace={key} key={key} />
```

File: packages/oscal-react-library/src/components/OSCALCatalogGroups.test.tsx (hunk 22 above)
Comment (kylelaker) [reply packages/oscal-react-library/src/components/OSCALCatalogGroups.test.tsx:56]: That's fine. I really dislike `test-data` but it exists and there's not much to do about that. If you move it, keep it in a TSX file.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 24 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:95]: Good catch!

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 23 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:63]: From looking at the many places in the spec where properties appear, it seems like they appear in a very similar manner, which makes sense to have properties appear as a row in a table. Although, we could instead call this `OSCALPropertyRow` instead if that makes more sense?

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 26 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:176]: Several of the tests will fair when I remove either this optional chaining operator or the one in `nist?.sort(byName)`

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 23 above)
Comment (kylelaker) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:63]: I guess my thought is that it shouldn't be a separate component -- it cannot be displayed outside `OSCALProperties` which is a table; so it's very dependent on that representation. But I guess since it's not exported, then it's likely fine.

File: packages/oscal-react-library/src/components/OSCALBackMatter.tsx (hunk 19 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALBackMatter.tsx]: Sorry, I should have mentioned I corrected this to what you were suggesting (`<NotSpecifiedTypography component="span">Resource</NotSpecifiedTypography>`)

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 26 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx]: I'm wrong. It is possible to have properties defined and only NIST properties, or have properties defined and only 3rd party properties. `nist?.sort(byName)` must have the optional chaining operator, but it should be fine to remove it from the place you mentioned.  With the 3rd party properties filter, it will handle if there are no 3rd parties defined, so I should be able to remove the optional chain operator here: `props?.sort(byName)`. Although the type declaration should also account for there being no NIST properties present - which is a case where there would be passed no properties passed to `OSCALProperties`.

File: packages/oscal-react-library/src/components/OSCALCatalogGroups.test.tsx (hunk 22 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALCatalogGroups.test.tsx:56]: I think I'll just plan to leave it where it is. Would it make sense to move more of our data to the file where we are testing it in the future?

Comment (tuckerzp) [reply packages/oscal-react-library/src/components/OSCALCatalogGroups.test.tsx:56]: I personally think having a huge list of "common" objects that we continuously add to makes testing a bit tricky. I think that testing the smallest unit possible with very minimum data makes our tests more understandable and makes it easier to test various edge cases. 

At the very least, we need to stop the pattern of missing data for tests, adding it to the already huge objects in `test-data` and making our tests more complex.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 27]
```diff
@@ -0,0 +1,183 @@
...
+  };
+
+  return (
+    <>
+      <OSCALPropertiesButton
+        size="small"
+        variant="outlined"
+        onClick={handleOpen}
+        aria-label={`${title} properties button`}
+        startIcon={<HomeIcon />}
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:149]: Instead of `Home` let's do https://mui.com/material-ui/material-icons/?selected=Construction

And then we can set 
```
title="Open Properties"
```

or something like that so that we can get _something_ on hover?

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+    setOpen(false);
+  };
+
+  return (
+    <>
+      <OSCALPropertiesButton
+        size="small"
+        variant="outlined"
+        onClick={handleOpen}
+        aria-label={`${title} properties button`}
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:148]: `title` may not be a string which won't work well in an `aria-label`. Also, I am not sure that "button" is appropriate here... We may want to use something like the other suggestion for the `title` attribute.

**Review by kylelaker — COMMENTED** [review]
Small iconography change and then I think I am good to move forward with merging.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx [hunk 29]
```diff
@@ -0,0 +1,183 @@
...
+                />
+              ))}
+          </TableBody>
+        </Table>
+      </TableContainer>
+    </>
+  );
+};
+
+interface OSCALPropertiesDialogProps {
```
Comment (kylelaker) [inline packages/oscal-react-library/src/components/OSCALProperties.tsx:107]:
```suggestion
export interface OSCALPropertiesDialogProps {
```

**Review by kylelaker — COMMENTED** [review]
Oh! We also need a storybook file for this! At least for the `OSCALProperties` dialog itself

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 27 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:149]: I much more favor that icon for the button instead. I was thinking more real estate in terms of property haha

`title` holds the control name, resource title, or name of document (for metadata) currently. I think this is useful for the modal to contain this info at the very least. Are you suggesting the button have the title "Open Properties"?

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 29 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:107]: Good point!

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 28 above)
Comment (Bronstrom) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:148]: I could just make this "open-properties" or should I be more specific? I was using `title` to differentiate them a little more, but I understand why that wouldn't work well.

File: packages/oscal-react-library/src/components/OSCALProperties.tsx (hunk 27 above)
Comment (kylelaker) [reply packages/oscal-react-library/src/components/OSCALProperties.tsx:149]: I am asking that the Button's `title` and `aria-label` attributes be set to "Open Properties"

**Review by kylelaker — APPROVED** [review]
Thanks for your work on this!

**Review by tuckerzp — APPROVED** [review]
